File: simple-linear-regression/simple_linear_regression.R
#Simple Linear Regression
dataset = read.csv('Salary_Data.csv')

#Splitting the dataset into training and test set (We need "caTools")
library(caTools)
set.seed(123) #The replica of random_state in python
split = sample.split(dataset$Salary, SplitRatio = 2/3)#Perform the split-- It doesnt matter what column you choose
training_set = subset(dataset, split == TRUE) #Apply the split on the dataset to generate the training
test_set = subset(dataset, split == FALSE)

#Feature scaling -- We need to exclude the columns with encoded categorical data (If any)
#This exclusion is done by specifying the unencoded data columns to be scaled
#training_set[, 2:3] = scale(training_set[, 2:3])
#test_set[, 2:3] = scale(test_set[, 2:3])

#We didnt need to scale because linear regression class takes care of that for us out of the box

#Fitting the linear regression to the training set -- Generate the model the program should use
regressor = lm(formula = Salary ~ YearsExperience,
               data = training_set)
#Above, you want to specify what columns are proportional in the dataset that you provide

#To properly view the details of the regressor, use the console to execute "summary(regressor)"

#Predicting the test set results 
y_pred = predict(regressor, newdata = test_set)

#Visualizing the training_set results
# install.packages('ggplot2')
library(ggplot2)
ggplot() +
  geom_point(aes(x = training_set$YearsExperience, y = training_set$Salary),
             colour = 'red') + #Plot the scatter points of the training observations
  geom_line(aes(x = training_set$YearsExperience, y = predict(regressor, newdata = training_set)),
            colour = 'blue') + #Draw a regression line based on the model applied on the training set
  ggtitle('Salary vs Experience (Training Set)') +
  xlab('Years Of Experience') +
  ylab('Salary')

#Visualizing the test_set results
ggplot() +
  geom_point(aes(x = test_set$YearsExperience, y = test_set$Salary),
             colour = 'red') +
  geom_line(aes(x = training_set$YearsExperience, y = predict(regressor, newdata = training_set)),
            colour = 'blue') + #Like in python, we dont need to use d test data to plot a new reg line as they would give line of same slope
  ggtitle('Salary vs Experience (Test Set)') +
  xlab('Years Of Experience') +
  ylab('Salary')
  

File: simple-linear-regression/simple_linear_regression.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jul 24 01:13:22 2018

@author: William
"""

#Importing libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#Importing the dataset
dataset = pd.read_csv('Salary_Data.csv')
X = dataset.iloc[:, :-1].values #Matrix of features i.e independent vars
y = dataset.iloc[:, 1].values


#Splitting dataset into training and test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 1/3, random_state = 0)

#Feature Scaling -- Used to make all the content of the dataset be of the same scale
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)"""

#Fitting the Simple Linear Regression to the training set
#This basically means making the program learn the correlation between the
#Independent and dependent variable.
#This is also how you make the model for the program
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

#Predict Test Set Result
y_pred = regressor.predict(X_test)

#Visualizing the results 
#The independent variable would be on the y-axis and vice-versa

#First plot the observation points of the real values
plt.scatter(X_train, y_train, color='red')
#Plot the regression line i.e the predictions
plt.plot(X_train, regressor.predict(X_train), color ='blue')
plt.title('Salary vs Experience (Training Set)')
plt.xlabel('Years Of Experience')
plt.ylabel('Salary')
plt.show()

#Now plot the observation points of the test values
plt.scatter(X_test, y_test, color='red')
#Below, it doesnt matter if u use x_test or train, the line remains the same
#The length could differ but the slope is the same
plt.plot(X_train, regressor.predict(X_train), color ='blue')
plt.title('Salary vs Experience (Test Set)')
plt.xlabel('Years Of Experience')
plt.ylabel('Salary')
plt.show()
